src/swap/utils/processSwapTransaction.ts

```diff
@@ -41,6 +50,39 @@ export async function processSwapTransaction({
       confirmations: 1,
     });
     setPendingTransaction(false);
+
+    // for the V2 API, we use Uniswap's `UniversalRouter`, which uses `Permit2` for ERC-20 approvals
+    // this adds an additional transaction/step to the swap process
+    // since we need to make an extra transaction to `Permit2` to allow the UniversalRouter to spend the approved funds
+    // this would typically be a (gasless) signature, but we're using a transaction here to allow batching for Smart Wallets
+    // read more: https://blog.uniswap.org/permit2-and-universal-router
+    if (!useAggregator) {
```

Comment:
I feel like this code, has a lot of important pieces, and there is only one unit-test checking on this.

I wonder how this could be break down in 2~3 unit-tests. Or even the code break down a bit more.

Reply:
the single unit test checks for all of the code/cases in this conditional
```
    expect(setPendingTransaction).toHaveBeenCalledTimes(6);
    expect(setPendingTransaction).toHaveBeenCalledWith(true);
    expect(setPendingTransaction).toHaveBeenCalledWith(false);
    expect(sendTransactionAsyncPermit2).toHaveBeenCalledTimes(3);
    expect(waitForTransactionReceipt).toHaveBeenCalledTimes(3);

    expect(setLoading).toHaveBeenCalledTimes(1);
    expect(setLoading).toHaveBeenCalledWith(true);
    expect(onSuccess).toHaveBeenCalledTimes(1);
    expect(onSuccess).toHaveBeenCalledWith({});
    expect(onStart).toHaveBeenCalledTimes(3);
    expect(onStart).toHaveBeenNthCalledWith(1, 'approveTxHash');
    expect(onStart).toHaveBeenNthCalledWith(2, 'permit2TxHash');
    expect(onStart).toHaveBeenNthCalledWith(3, 'txHash');
```

there can be an argument to decompose into two separate functions - `processSwapApprovals` and `processSwapTransaction`, both of which are under `processSwap` parent function. I can take this action item as part of the Smart Wallet integration.